main/src/ca/uwaterloo/flix/language/phase/Inliner.scala

```diff
@@ -34,20 +40,350 @@ import ca.uwaterloo.flix.util.Validation._
+      case ApplyTail(_, _, _, _, _) => None
+      case ApplyHook(_, _, _, _) => None
+      case Apply(_, _, _, _) => None
+      case Unary(_, _, _, _) => None
+      case Binary(_, _, _, _, _) => None
+      case IfThenElse(_, _, _, _, _) => None
+      case Let(_, _, _, _, _) => None
+      case LetRec(_, _, _, _, _) => None
+      case Is(_, _, _, _) => None
+      case Tag(_, _, _, _, _) => None
+      case Untag(_, _, _, _, _) => None
+      case Index(_, _, _, _) => None
+      case Tuple(_, _, _) => None
+      case Existential(_, _, _) => None
+      case Universal(_, _, _) => None
+      case NativeConstructor(_, _, _, _) => None
+      case NativeField(_, _, _) => None
+      case NativeMethod(_, _, _, _) => None
+      case UserError(_, _) => None
+      case MatchError(_, _) => None
+      case SwitchError(_, _) => None
+    }
+    AstVisitor.visitExpressionWith(f, exp)
+  }
 
-    // TODO: Implement.
+  /**
+    * Either performs some high-level inlining using `scores`
+    * or returns None to ask traversal to inline inside `exp`
+    */
+  def inlineVisitor(definitions: Map[DefnSym, Constant], scores: Map[DefnSym, Int], exp: Expression)(implicit genSym: GenSym): Option[Expression] =
+  {
+    /**
+      * Returns args and the symbols they should be bound to in required order for substitution and let binding
+      * @param args List[Expression] of arguments
+      * @return (revArgs, argSymbols) where argSymbols is arguments with all arguments replaced with appropriate symbols
+      *         and revArgs are non-var arguments in reverse order
+      */
+    def argsToSym(args: List[Expression]): (List[(Expression,VarSym)], List[VarSym]) =
+    {
+      /**
+        * @param letAcc Arguments and symbols they need to be bound to in reverse order
+        * @param symAcc Symbols to use for inlining in reverse order
+        * @param args Arguments to replace with variables if not variables
+        */
+      @tailrec def reverseArgVars(letAcc: List[(Expression,VarSym)], symAcc: List[VarSym], args: List[Expression]):
+      (List[(Expression,VarSym)], List[VarSym]) =
+      {
+        args match {
+          case ::(head, tl) =>
+            head match {
+              case v: Var =>
+                reverseArgVars(letAcc, v.sym::symAcc, tl)
+              case _ =>
+                val s = Symbol.freshVarSym("inlnrL")
+                reverseArgVars((head, s)::letAcc, s::symAcc, tl)
+            }
+          case Nil =>
+            (letAcc, symAcc.reverse)
+        }
+      }
+
+      reverseArgVars(Nil, Nil, args)
+    }
+
+    @tailrec def letBindAndInlineArgs(revArgs: List[(Expression,VarSym)], expr: Expression): Expression =
+    {
+      revArgs match {
+        case ::(head, tl) =>
+          val letTerm = inlineExpr(definitions,scores,head._1)
+          val letSym = head._2
+          letBindAndInlineArgs(tl, Let(letSym,letTerm,expr,expr.tpe,letTerm.loc))
+        case Nil =>
+          expr
+      }
+    }
```

Comment:
Also, if the idea is that in the list `Expr, VarSym` that VarSym is bound to `Expr` then it would better to swap the order `VarSym, Expr` which follows the intuition `x = expr` better.